overlord/snapstate/snapmgr.go

```diff
@@ -908,14 +927,23 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	if err != nil {
 		return err
 	}
+	if oldInfo != nil {
+		// TODO: move this to its own task
+		m.backend.ClearTrashedData(oldInfo)
+	}
 
 	// save for undoLinkSnap
 	t.Set("old-trymode", oldTryMode)
 	t.Set("old-devmode", oldDevMode)
 	t.Set("old-jailmode", oldJailMode)
 	t.Set("old-channel", oldChannel)
 	t.Set("old-current", oldCurrent)
-	t.Set("had-candidate", hadCandidate)
+	if ss.HasABA() {
+		t.Set("old-candidate-index", oldCandidateIndex)
+	} else {
+		//migration
+		t.Set("had-candidate", oldCandidateIndex < 0)
```

Comment:
the old code afaiu had:  oldCandidateIndex < 0 => had-candidate = false  so this looks wrong and isn't probably tested enough